apsi-enrollment-be/src/main/java/edu/pw/apsienrollment/common/db/SearchSpecification.java

```diff
@@ -0,0 +1,52 @@
+package edu.pw.apsienrollment.common.db;
+
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import org.springframework.data.jpa.domain.Specification;
+
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.Path;
+import javax.persistence.criteria.Predicate;
+import javax.persistence.criteria.Root;
+import java.util.Arrays;
+import java.util.Collection;
+
+@RequiredArgsConstructor
+public abstract class SearchSpecification<S> implements Specification<S> {
+    protected final Collection<SearchCriteria> searchCriteria;
+
+    protected Predicate toCommonSearchPredicate(SearchCriteria criteria, Root<S> root, CriteriaBuilder criteriaBuilder) {
+        Path<String> path = resolvePath(criteria.getKey(), root);
+        if (criteria.getOperation().equalsIgnoreCase(">")) {
+            return criteriaBuilder.greaterThan(
+                    path, criteria.getValue().toString());
+        } else if (criteria.getOperation().equalsIgnoreCase("<")) {
+            return criteriaBuilder.lessThan(
+                    path, criteria.getValue().toString());
+        } else if (criteria.getOperation().equalsIgnoreCase(">=")) {
+            return criteriaBuilder.greaterThanOrEqualTo(
+                    path, criteria.getValue().toString());
+        } else if (criteria.getOperation().equalsIgnoreCase("<=")) {
+            return criteriaBuilder.lessThanOrEqualTo(
+                    path, criteria.getValue().toString());
+        } else if (criteria.getOperation().equalsIgnoreCase(":")) {
+            if (path.getJavaType() == String.class) {
+                return criteriaBuilder.like(
+                        criteriaBuilder.function("LOWER", String.class, path),
+                        "%" + criteria.getValue().toString().toLowerCase() + "%");
+            } else {
+                return criteriaBuilder.equal(path, criteria.getValue());
+            }
```

Comment:
Code redundancy could be avoided (and readability improved) by using a map string => method, however I don't know if this practice is accepted in java/spring

Reply:
This could be done if only all those methods weren't called on object passed through argument. As you can see it's always a call to criteriaBuilder object.